src/boot/efi/vmm.c

```diff
@@ -241,21 +241,15 @@ static const SmbiosHeader *get_smbios_table(uint8_t type, uint64_t *ret_size_lef
                 size -= header->length;
                 p += header->length;
 
-                /* Skip over string table. */
-                for (;;) {
-                        const uint8_t *e = memchr(p, 0, size);
-                        if (!e)
-                                return NULL;
-
-                        if (e == p) {/* Double NUL byte means we've reached the end of the string table. */
-                                p++;
-                                size--;
-                                break;
-                        }
-
-                        size -= e + 1 - p;
-                        p = e + 1;
+                /* Skip chars until a double NUL */
+                while (!(p[0] == '\0' && p[1] == '\0')) {
```

Comment:
so this is not right:

what if the size left is actually zero or 1? then you derefence invalid memory.

you need to take the size into account first before dereferencing, which the code above did.

also, memchr() is typically a myriad times faster than manual loops because off sse implementations made available by compilers. I think it would be good to stick to that.

Reply:
@poettering 

If `p[0]` is a \0 byte, either another terminating \0 byte comes after it or the first byte of the next string, thus the deref of `p[1]` on the condition of `p[0] == '\0'` is valid under the precondition of spec compliance.

Of course if we assume the firmware can produce garbage SMBIOS tables, then yeah it should be guarded against.

Reply:
yeah, i would be defensive here. formatting tables is hard, so far we tried to deal gracefully with truncated smbios tables. the linux kernel is really defensive on this too, and this code runs with equal privileges, hence i think we should be equally careful

Reply:
Yeah makes sense, I'll try to find some time to fix this and the rest of the things you have pointed out.